### DISCUSSION TITLE
VSCode `Format on Save` doesn't work with Darker

### DISCUSSION THREAD

**Discussion by agandra** [akaihola/darker#318] [Q&A]
Hey, I have setup my black and darker inside a dev container using the Remote Container plugin on VSCode.  However, when I make changes the formatting is not reflected in my codebase.  Nothing happens on save even though darker is running.  It worked on the entire file when I pointed it at black instead of darker.

Here is my dev container file:

```
{
    "name": "Docker Test",
    "build": {
        "dockerfile": "../Dockerfile",
    },
    // Set *default* container specific settings.json values on container create.
    "settings": {
        "python.pythonPath": "/usr/local/bin/python",
        "python.languageServer": "Pylance",
        "python.linting.enabled": true,
        "python.formatting.blackPath": "/usr/local/bin/darker",
        "python.formatting.provider": "black",
        "editor.formatOnSave": true,
        "python.formatting.blackArgs": [
            "--diff"
        ],
    },
    // Add the IDs of extensions you want installed when the container is created.
    "extensions": [
        "ms-python.python",
        "ms-python.vscode-pylance"
    ],
    "postCreateCommand": "pip install -r requirements-dev.txt"
}
```

When I look at my output for python when i make changes this is what appears:

```
> /usr/local/bin/darker --diff --diff --quiet ./main_app/views.py
cwd: .
> /usr/local/bin/darker --diff --diff --quiet ./main_app/views.py.db72b8e8942140e8941050de60babb6d.tmp
cwd: .
> /usr/local/bin/darker --diff --diff --quiet ./main_app/views.py.db72b8e8942140e8941050de60babb6d.tmp
cwd: .
> /usr/local/bin/darker --diff --diff --quiet ./main_app/views.py.db72b8e8942140e8941050de60babb6d.tmp
cwd: .
> /usr/local/bin/darker --diff --diff --quiet ./main_app/views.py.db72b8e8942140e8941050de60babb6d.tmp
cwd: .
> /usr/local/bin/darker --diff --diff --quiet ./main_app/views.py.db72b8e8942140e8941050de60babb6d.tmp
cwd: .
```

If i run in my terminal manually: 

> /usr/local/bin/darker --diff --diff ./main_app/views.py

I see the output being the file formatted correctly.  However, there are is no change happening within the actual code.

![Screen Shot 2022-02-15 at 2 01 01 PM](https://user-images.githubusercontent.com/1072647/154130588-bd62db02-75d9-43e4-a9c7-778263ac551c.png)

Thanks!

**Comment by akaihola** [comment]
Hi @agandra, thanks for your report!

Which Darker version are you running? Support for VSCode `formatOnSave` was added in version 1.3.2.

Does Darker apply formatting changes correctly if you run VSCode with a local repository and a local Darker instead of using the Remote Container plugin?

Does it work if you explicitly run the `Format Document` command in VSCode?

**Comment by agandra** [comment]
Pip freeze returns that i'm on darker 1.4.0.

I run all my python code through docker containers, but will try to get python working on my local machine to install locally and see if darker works.

If i run Format Document manually it formats the entire document correctly.  So it appears that for some reason it runs format on save but the actual file doesn't format.  But when I do it explicitly it works.

**Comment by akaihola** [comment]
>If i run Format Document manually it formats the entire document

If you've set up Darker disguising as Black correctly (as it seems from the configuration you pasted), it shouldn't format the entire document, but only the modified lines. So maybe something wrong there as well?

**Comment by agandra** [comment]
Sorry, correction on what I previously said.  It only formats the diff correctly, does not effect unmodified code.  So I think its working properly, it just does not happen on save.

**Comment by akaihola** [comment]
Ok thanks for the clarification.

And sorry for mixing things up: Last week we finally merged another fix for `formatOnSave` for VSCode in #255. This will be released as part of version 1.4.1 once we get #287 done, reviewed and merged (help accepted :).

(There was only a partial fix in #215 for version 1.3.2.)

You can test the fix in #255 before 1.4.1 is released by installing Darker from GitHub `master`.

**Comment by mayk0gan** [comment]
In the meanwhile, you can still make the formatter work on save by clicking on save twice.
(Yeah, it's pretty annoying but it works until @akaihola's important fix)

**Comment by akaihola** [comment]
@agandra Darker 1.4.1 has now been released and the fix in #255 is there. Thanks again for your report, and let us know if there are still VSCode (or other) issues!

**Reply by tobiasdiez** [reply]
For me this is still not working correctly. When I save a file, darker is correctly run on the tmp file, but no changes are performed. Only after saving a few times (normally between 2 or 3), the format changes are applied.

**Reply by akaihola** [reply]
Thanks @tobiasdiez for pointing that out. Assuming all works correctly when using Black as a formatter instead of Darker, we really need to find out what the significant difference between the behavior of Black and Darker is. Do you have any ideas for how to do that?

**Reply by tobiasdiez** [reply]
I don't have much experience with how black and darker work, so I can report from the user's perspective. Would it for example be possible to print the diff that black and darker are producing on save to compare the output? A shot in the blue: maybe darker has troubles finding the "original" version? I'm happy to investigate this further, but need assistance of what cli args to use and what to look out for etc.

**Reply by akaihola** [reply]
Well, if you can see the exact command line VSCode is using to call Black or Darker, you can issue exactly the same command manually and take a look at the result.

To be absolutely sure about the command line, you could point `python.formatting.blackPath` in VSCode to a custom script like the one below instead of Black or Darker:
```bash
#!/usr/bin/env bash

echo "$@" >/tmp/vscode-black-command-line.txt
```

You could then save a Python file in VSCode and verify that `/tmp/vscode-black-command-line.txt` gets created. Then create a tmp file, run both Darker and Black on the tmp file manually with the arguments of that command line, and compare the outputs. Also pay attention to whether the tmp file was modified or not.

Hope this helps!

**Comment by mayk0gan** [comment]
Unfortunately, it still doesn't work - but now (as a patch of course), sometimes it works when pressing save 3 times instead (or even more)
From VSCode's logs:

```
Formatting with black failed.
Error: Invalid patch mode 'd' in: ...py.d4fc638eda7bd41ef314b24cb7d84214.tmp:14: error: Function is missing a type annotation  [no-untyped-def]
``` 

where the `...` part is my file path, with missing chars at the beginning of it.

In the second save, less chars a missing (only the the first '/').

**Comment by akaihola** [comment]
>`error: Function is missing a type annotation  [no-untyped-def]`

@mayk0gan, this looks like you've enabled the `mypy` linter either in your VSCode Black configuration or the Darker configuration section in `pyproject.toml`. Darker appends `mypy` output to the Black reformatting diff, and VSCode doesn't expect that.

I guess we could add a warning or a descriptive error message whenever Darker encounters a `.py.<hash>.tmp` file and has any `--lint`/`-L` options enabled (see #297).

**Comment by akaihola** [comment]
@agandra and @tobiasdiez, do you still have the issue of Darker not reformatting on save in VSCode when using Darker 1.4.1 or newer and making sure no linters are run (either with `-L`/`--lint` on the command line or `lint = ...` in `pyproject.toml`)?
